File: samples/quickstart/python/09_capability_discovery.py
"""
09 - Capability Discovery
==========================

Shows how to explore available capabilities without memorizing
full dotted paths. The capabilities API lets you list, resolve,
search, and visualize the capability hierarchy.

No API keys needed -- this is pure introspection.
"""

from __future__ import annotations

import json

import modelmesh


def main():
    # 1. List all registered capability aliases
    print("=== All Capabilities ===")
    caps = modelmesh.capabilities.list_all()
    for cap in caps:
        full_path = modelmesh.capabilities.resolve(cap)
        print(f"  {cap:25s} -> {full_path}")

    # 2. Resolve alias to full dotted path
    print("\n=== Resolve Alias ===")
    alias = "chat-completion"
    path = modelmesh.capabilities.resolve(alias)
    print(f"  '{alias}' -> '{path}'")

    # Already-dotted paths pass through unchanged
    print(f"  '{path}' -> '{modelmesh.capabilities.resolve(path)}'")

    # Unknown aliases return unchanged
    unknown = "my-custom-capability"
    print(f"  '{unknown}' -> '{modelmesh.capabilities.resolve(unknown)}'")

    # 3. Search by keyword
    print("\n=== Search ===")
    for keyword in ["text", "audio", "image"]:
        matches = modelmesh.capabilities.search(keyword)
        print(f"  '{keyword}': {matches}")

    # 4. Full hierarchy tree
    print("\n=== Capability Tree ===")
    tree = modelmesh.capabilities.tree()
    print(json.dumps(tree, indent=2))


if __name__ == "__main__":
    main()

File: samples/quickstart/typescript/09_capabilityDiscovery.ts
/**
 * 09 - Capability Discovery
 * ==========================
 *
 * Shows how to explore available capabilities without memorizing
 * full dotted paths. The capabilities API lets you list, resolve,
 * search, and visualize the capability hierarchy.
 *
 * No API keys needed — this is pure introspection.
 */

import * as capabilities from '@nistrapa/modelmesh-core/capabilities';

function main(): void {
  // 1. List all registered capability aliases
  console.log('=== All Capabilities ===');
  const caps = capabilities.listAll();
  for (const cap of caps) {
    const fullPath = capabilities.resolve(cap);
    console.log(`  ${cap.padEnd(25)} → ${fullPath}`);
  }

  // 2. Resolve alias to full dotted path
  console.log('\n=== Resolve Alias ===');
  const alias = 'chat-completion';
  const path = capabilities.resolve(alias);
  console.log(`  '${alias}' → '${path}'`);

  // Already-dotted paths pass through unchanged
  console.log(`  '${path}' → '${capabilities.resolve(path)}'`);

  // Unknown aliases return unchanged
  const unknown = 'my-custom-capability';
  console.log(`  '${unknown}' → '${capabilities.resolve(unknown)}'`);

  // 3. Search by keyword
  console.log('\n=== Search ===');
  for (const keyword of ['text', 'audio', 'image']) {
    const matches = capabilities.search(keyword);
    console.log(`  '${keyword}': [${matches.join(', ')}]`);
  }

  // 4. Full hierarchy tree
  console.log('\n=== Capability Tree ===');
  const tree = capabilities.tree();
  console.log(JSON.stringify(tree, null, 2));
}

main();
